monero-harness/src/lib.rs

```diff
@@ -44,219 +45,280 @@ const BLOCK_TIME_SECS: u64 = 1;
 /// Poll interval when checking if the wallet has synced with monerod.
 const WAIT_WALLET_SYNC_MILLIS: u64 = 1000;
 
-/// Wallet sub-account indices.
-const ACCOUNT_INDEX_PRIMARY: u32 = 0;
-
-#[derive(Copy, Clone, Debug)]
+#[derive(Clone, Debug)]
 pub struct Monero {
-    monerod_rpc_port: u16,
-    miner_wallet_rpc_port: u16,
-    alice_wallet_rpc_port: u16,
-    bob_wallet_rpc_port: u16,
+    monerod: Monerod,
+    wallets: Vec<MoneroWalletRpc>,
+    container_prefix: String,
 }
-
 impl<'c> Monero {
-    /// Starts a new regtest monero container.
-    pub fn new(cli: &'c Cli) -> Result<(Self, Container<'c, Cli, image::Monero>)> {
-        let monerod_rpc_port: u16 =
-            port_check::free_local_port().ok_or_else(|| anyhow!("Could not retrieve free port"))?;
-        let miner_wallet_rpc_port: u16 =
-            port_check::free_local_port().ok_or_else(|| anyhow!("Could not retrieve free port"))?;
-        let alice_wallet_rpc_port: u16 =
-            port_check::free_local_port().ok_or_else(|| anyhow!("Could not retrieve free port"))?;
-        let bob_wallet_rpc_port: u16 =
-            port_check::free_local_port().ok_or_else(|| anyhow!("Could not retrieve free port"))?;
-
-        let image = image::Monero::default()
-            .with_mapped_port(Port {
-                local: monerod_rpc_port,
-                internal: MONEROD_RPC_PORT,
-            })
-            .with_mapped_port(Port {
-                local: miner_wallet_rpc_port,
-                internal: MINER_WALLET_RPC_PORT,
-            })
-            .with_wallet("miner", MINER_WALLET_RPC_PORT)
-            .with_mapped_port(Port {
-                local: alice_wallet_rpc_port,
-                internal: ALICE_WALLET_RPC_PORT,
-            })
-            .with_wallet("alice", ALICE_WALLET_RPC_PORT)
-            .with_mapped_port(Port {
-                local: bob_wallet_rpc_port,
-                internal: BOB_WALLET_RPC_PORT,
-            })
-            .with_wallet("bob", BOB_WALLET_RPC_PORT);
-
-        println!("running image ...");
-        let docker = cli.run(image);
-        println!("image ran");
+    /// Starts a new regtest monero container setup consisting out of 1 monerod
+    /// node and n wallets. The containers will be prefixed with
+    /// `container_prefix` if provided. There will be 1 miner wallet started
+    /// automatically. Default monerod container name will be: `monerod`
+    /// Default miner wallet container name will be: `miner`
+    /// Default network will be: `monero`
+    pub async fn new(
+        cli: &'c Cli,
+        container_prefix: Option<String>,
+        network_prefix: Option<String>,
```

Comment:
I simplified it to one prefix now. The problem is that you can only have 1 container at the time with a particular name. Running tests in parallel is then not possible, hence each test can define its own prefix.

Reply:
How about a random prefix?

Reply:
When debugging I found it easier to know which container to look for.

Reply:
I get that, but for automated tests on CI (for example), we're gonna default to passing in a random prefix anyway (assuming that if you pass `None` it can fail because we run tests in parallel). 

Alternatively, you could default to a random prefix if `None` is passed.

Reply:
Done, if not provided a random prefix will be generated.